README.md
# OCIO STI Concept Tagging Service

An API for exposing models created with [STI concept training](https://github.com/nasa/concept-tagging-training). This project was written about [here](https://strategy.data.gov/proof-points/2019/05/28/improving-data-access-and-data-management-artificial-intelligence-generated-metadata-tags-at-nasa/) for the Federal Data Strategy Incubator Project. 

### What is Concept Tagging
By concept tagging, we mean you can supply text, for example:
`Volcanic activity, or volcanism, has played a significant role in the geologic evolution of Mars.[2] Scientists have known since the Mariner 9 mission in 1972 that volcanic features cover large portions of the Martian surface.` and get back predicted keywords, like `volcanology, mars surface, and structural properties`, as well as topics, like `space sciences, geosciences`, from a standardized list of several thousand NASA concepts with a probability score for each prediction.

## Index
1. [Using Endpoint](#using-endpoint)
    1. [Request](#request)
    2. [Response](#response)
2. [Running Your Own Instance](#running-your-own-instance)
    1. [Installation](#installation)
        1. [Pull Docker Image](#pull-docker-image)
        2. [Build Docker Image](#build-docker-image)
        3. [With Local Python](#with-local-python)
    2. [Download Models](#downloading-models)
    3. [Running Service](#running-service)
        1. [Using Docker](#using-docker)
        2. [Using Local Python](#using-local-python)

## Using Endpoint
### Request
The endpoint accepts a few fields, shown in this example:
```json
{
    "text": [
        "Astronauts go on space walks.",
        "Basalt rocks and minerals are on earth."
    ], 
    "probability_threshold":"0.5",
    "topic_threshold":"0.9", 
    "request_id":"example_id10"
}
```
- **text** *(string or list of strings)* -- The text(s) to be tagged.
- **probability_threshold** *(float in [0, 1])* -- a threshold under which a concept tag will not be returned by the API. For example, if the threshold is set to 0.8 and a concept only scores 0.5, the concept will be omitted from the response. Setting to 1 will yield no results. Setting to 0 will yield all of the classifiers and their scores, no matter how low.
- **topic_threshold** *(float in [0, 1])* -- A probability threshold for categories. If a category falls under this threshold, its respective suite of models will not be utilized for prediction. If you set this value to 1, only the generalized concept models will be used for tagging, yielding significant speed gains.
- **request_id** *(string)* -- an optional ID for your request.  

You might send this request using curl. In the command below:
1. Substitute `example_payload_multiple.json` with the path to your json request.
2. Substitute `http://0.0.0.0:5000/` with the address of the API instance.
```
curl -X POST -H "Content-Type: application/json" -d @example_payload_multiple.json http://0.0.0.0:5000/findterms/
```
### Response
You will then receive a response like that [here](docs/multiple_response.json). In the `payload`, you will see multiple fields, including:
- **features** -- words and phrases directly extracted from the document. 
- **sti_keywords** -- concepts and their prediction scores. 
- **topic_probability** -- model scores for all of the categories.

## Running Your Own Instance
### Installation
For most people, the simplest installation entails [building the docker image](#build-docker-image), [downloading the models](#downloading-models), and [running the docker container](#using-docker).


#### Build Docker Image
First, clone this repository and enter its root.
Now, you can build the image with:
```
docker build -t concept_tagging_api:example .
```
\* Developers should look at the `make build` command in the [Makefile](Makefile). It has an automated process for tagging the image with useful metadata.

#### With Local Python
\* tested with python:3.7  
First, clone this repository and enter its root.  
Now, create a virtual environment. For example, using [venv](https://docs.python.org/3/library/venv.html):
```
python -m venv venv
source venv/bin/activate
```
Now install the requirements with:
```
make requirements
```

### Downloading Models
Then, you need to download the machine learning models upon which the service relies. 

You can find zipped file which contains all of the models [here](https://data.nasa.gov/docs/datasets/public/concept_tagging_models/10_23_2019.zip). Now, to get the models in the right place and unzip:
```bash
mkdir models
mv <YOUR_ZIPPED_MODELS_NAME>.zip models
cd models
unzip <YOUR_ZIPPED_MODELS_NAME>.zip
```
Alternatively, the models can also be downloaded from data.nasa.gov where they are named <a href='https://data.nasa.gov/Software/STI-Tagging-Models/jd6d-mr3p'>STI Tagging Models</a>. However, they download slower from that location.

### Running Service

#### Using Docker
With the docker image and model files in place, you can now run the service with a simple docker command. In the below command be sure to:
 1. Substitute `concept_tagging_api:example` for the name of your image.
 2. Substitute `$(pwd)/models/10_23_2019` to the path to your models directory. 
 3. Substitute `5001` with the port on your local machine from which you wish to access the API.
```
docker run -it \
    -p 5001:5000 \
    -v $(pwd)/models/10_23_2019:/home/service/models/experiment \
    concept_tagging_api:example
```

Note that you you may experience permission errors when you start the container. To resolve this issue, set the user and group of your `models` directory to 999. This is the uid for the user 

**optional**
The entrypoint to the docker image is [gunicorn](https://docs.gunicorn.org/en/stable/index.html), a python WSGI HTTP Server which runs our flask app. You can optionally pass additionally arguments to gunicorn. For example:
```bash
docker run -it \
    -p 5001:5000 \
    -v $(pwd)/models/10_23_2019:/home/service/models/experiment \
    concept_tagging_api:example --timeout 9000 
```
See [here](https://docs.gunicorn.org/en/stable/design.html#async-workers) for more information about design considerations for these gunicorn settings.

#### Pitfalls & Gotchas to Remeber
- If you run this on a cloud service and run an upgrade on everything out of date for security reasons, you may need to run `sudo service docker stop`
and then `sudo service docker start` to get docker going again. You'll also have to find the docker container that you had last running and restart it.
- If you run the docker container as described above, remember to try the URL of your service with the proper port at the end of the URL. 

#### Using Local Python
With the requirements installed and the model files in place, you can now run the service with python locally. 
In the command below, substitute `models/test` with the path to your models directory. For example, if you followed the example from [With Bucket Access](#with-bucket-access), it will be `models/10_23_2019`.
```
export MODELS_DIR=models/test; \
python service/app.py
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

Releases page is <a href="https://github.com/nasa/concept-tagging-api/releases">here</a>.

## [Unreleased]

## [v5.0.12] - 2020-06-16
#### Fixed:
Models are now loaded before the API starts.

#### Added:
Added option to preload the models or load them for each request.

## [v5.0.11] - 2020-06-16
#### Fixed:
Fixed various links to point to the open source code respositories instead of internal NASA repositories.

#### Added:
 - Make command to download models from data.nasa.gov.
 - .ebsextensions command for downloading models for elastic beanstalk deployment


## [v5.0.9-open_source_release] - 2020-06-10
#### Added:
Original open-source release of this repository on github.com/nasa after having received SRA (software release authority) approval.



# Guidelines for ChangeLog Entries

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

### Guiding Principles
- Changelogs are for humans, not machines.
- There should be an entry for every single version.
- The same types of changes should be grouped.
- Versions and sections should be linkable.
- The latest version comes first.
- The release date of each version is displayed.

### All Entries Sould be Under One of These Types of changes
- <b>Added</b> for new features.
- <b>Changed</b> for changes in existing functionality.
- <b>Deprecated</b> for soon-to-be removed features.
- <b>Removed</b> for now removed features.
- <b>Fixed</b> for any bug fixes.
- <b>Security</b> in case of vulnerabilities.

Google technical writer Sarah Maddox <a href="https://ffeathers.wordpress.com/2017/08/19/how-to-write-release-notes/">gave the following advice</a> about release notes:
`“The most important function of release notes is to let customers know that something has changed in the product, particularly when that something may affect the way the customer uses the product.”`</a>


tests/context.py
import os
import sys

sys.path.insert(
    0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../service/"))
)

import app


tests/test_app.py
import unittest
from .context import app
from app import app, SERVICE_ROOT_PATH, FIND_TERMS_METHOD_NAME
from tempfile import TemporaryDirectory
from dsconcept.get_metrics import StubBestEstimator, HierarchicalClassifier
from sklearn.feature_extraction import DictVectorizer
import json


class MyTestCase(unittest.TestCase):
    def setUp(self):
        self.d = TemporaryDirectory()
        cat_clfs = [
            {"best_estimator_": StubBestEstimator(), "concept": "physics"},
            {"best_estimator_": StubBestEstimator(), "concept": "video games"},
        ]
        kwd_clfs = {
            ("physics", "gamma ray"): StubBestEstimator(),
            ("video games", "minecraft"): StubBestEstimator(),
            ("video games", "kerbal space program"): StubBestEstimator(),
            ("", "minecraft"): StubBestEstimator(),
            ("", "gamma ray"): StubBestEstimator(),
            ("", "penguins"): StubBestEstimator(),
        }
        v = DictVectorizer()
        d = [{"astronauts": 1, "astronomy": 1}, {"space": 1, "basalt": 1}]
        v.fit(d)
        app.config["HIER_CLF"] = HierarchicalClassifier(cat_clfs, kwd_clfs)
        app.config["HIER_CLF"].vectorizer = v
        app.config["CAT_RAW2LEMMA"] = {
            "PHYSICS": "physics",
            "VIDEO GAMES": "video games",
        }
        app.config["CAT_LEMMA2RAW"] = {
            v: k for k, v in app.config["CAT_RAW2LEMMA"].items()
        }
        app.config["KWD_RAW2LEMMA"] = {
            "MINECRAFT": "minecraft",
            "GAMMA RAY": "gamma ray",
            "KERBAL SPACE PROGRAM": "kerbal space program",
            "PENGUINS": "penguins",
        }
        app.config["KWD_LEMMA2RAW"] = {
            v: k for k, v in app.config["KWD_RAW2LEMMA"].items()
        }
        app.config["WEIGHTS"] = {
            "NOUN": 1,
            "NOUN_CHUNK": 1,
            "ENT": 1,
            "PROPN": 1,
            "ACRONYM": 1,
        }
        self.app = app.test_client()

    def test_home(self):
        response = self.app.get(f"{SERVICE_ROOT_PATH}/")
        self.assertEqual(response.status_code, 200)

    def test_find_terms(self):
        data = {
            "text": [
                "Astronauts go on space walks.",
                "Basalt rocks and minerals are on earth.",
            ],
            "probability_threshold": "0.5",
            "topic_threshold": "0.9",
            "request_id": "example_id10",
        }
        response = self.app.post(
            f"{SERVICE_ROOT_PATH}/{FIND_TERMS_METHOD_NAME}/",
            data=json.dumps(data),
            follow_redirects=True,
        )
        self.assertEqual(response.status_code, 200)
        r_data = json.loads(response.data)
        for rt in ["features", "sti_keywords", "topic_probabilities"]:
            self.assertEqual(len(r_data["payload"][rt]), 2)

    def test_find_terms_single(self):
        data = {
            "text": "Astronauts go on space walks.",
            "probability_threshold": "0.5",
            "topic_threshold": "0.9",
            "request_id": "example_id10",
        }
        response = self.app.post(
            f"{SERVICE_ROOT_PATH}/{FIND_TERMS_METHOD_NAME}/",
            data=json.dumps(data),
            follow_redirects=True,
        )
        self.assertEqual(response.status_code, 200)
        r_data = json.loads(response.data)
        for rt in ["features", "sti_keywords", "topic_probabilities"]:
            self.assertEqual(len(r_data["payload"][rt]), 1)


if __name__ == "__main__":
    unittest.main()


src/get_tag_names.py
import argparse
import logging
from pathlib import Path
import json

import joblib
from tqdm import tqdm


logging.basicConfig(level=logging.INFO)
LOG = logging.getLogger(__name__)
LOG.setLevel(logging.INFO)

def get_dir_cons(mdir):
    cons = []
    for p in tqdm(mdir.iterdir()):
        c = joblib.load(p)['concept']
        cons.append(c)
    return cons


def main(model_dir, text_dir):
    LOG.info(f'Reading models from {model_dir}.')

    d = {
        'cat': model_dir / 'categories/models', 
        'kwd': model_dir / 'keywords/models/topic_',
    }
    for t, p in d.items():
        raw2lemma_loc = model_dir / f'{t}_raw2lemma.json'
        with open(raw2lemma_loc, 'r') as f0:
            raw2lemma = json.load(f0)
        lemma2raw = {v: k for k, v in raw2lemma.items()}
        tags = get_dir_cons(p)
        raw_tags = [lemma2raw[c] for c in tags]
        out_p = model_dir / f"{t}_list.txt"
        LOG.info(f'Writing {len(tags)} {t}s to {out_p}.')
        with open(out_p, 'w') as f0:
            for l in raw_tags:
                f0.write(l)
                f0.write('\n')


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description = 'Make text files of all keyword tags from models.')
    parser.add_argument('--model_dir', help='input txt file', type=Path)
    parser.add_argument('--text_dir', help='input txt file', type=Path)
    args = parser.parse_args()
    main(args.model_dir, args.text_dir)


service/app.py
import json
import logging
import os
from pathlib import Path

import dsconcept.get_metrics as gm
import joblib
import yaml
from flask import request, jsonify, render_template, Flask

logging.basicConfig(level=logging.INFO)
LOG = logging.getLogger(__name__)
LOG.setLevel(logging.INFO)

app = Flask(__name__)
app.config["MODELS_DIR"] = os.environ["MODELS_DIR"]
LOG.info(f"Using MODELS_DIR \"{app.config['MODELS_DIR']}\"")

FIND_TERMS_METHOD_NAME = "findterms"
ALLOWED_FIND_TERMS_FIELDS = [
    "text",
    "probability_threshold",
    "topic_threshold",
    "request_id",
]

SERVICE_VERSION = os.environ["VERSION"] if "VERSION" in os.environ else "unspecified"

if "PRELOAD" in os.environ:
    if os.environ["PRELOAD"].lower().strip() == "true":
        PRELOAD = True
        LOG.info("Preloading the models.")
    else:
        PRELOAD = False
else:
    LOG.info("Not PRELOAD env variable found. Default to PRELOAD=True")
    PRELOAD = True

INTERFACE_VERSION = "2.0.0"
SERVICE_ROOT_PATH = ""

app.config.update(
    dict(
        CATEGORIES_DIR=f"{app.config['MODELS_DIR']}/categories/models",
        CONCEPTS_DIR=f"{app.config['MODELS_DIR']}/keywords/models",
        IN_KWD_RAW2LEMMA=f"{app.config['MODELS_DIR']}/kwd_raw2lemma.json",
        IN_CAT_RAW2LEMMA=f"{app.config['MODELS_DIR']}/cat_raw2lemma.json",
        IN_VECTORIZER=f"{app.config['MODELS_DIR']}/vectorizer.jbl",
        STI_CONFIG=f"{app.config['MODELS_DIR']}/config.yml",
        PRELOAD=PRELOAD,  # TODO: visible configuration for preload option
        LOC_DICT=Path("loc_dict.jbl"),
    )
)


def load_models_wrapper():
    """
    If not preloading and mapping from (topic, keyword) to model paths already exists,
    just load it directly. Otherwise, load in all models and create the file.
    """
    if (not app.config['PRELOAD']) and (app.config['LOC_DICT'].exists()):
        cd = joblib.load(app.config['LOC_DICT'])
    elif not app.config['PRELOAD']:
        cd = gm.load_concept_models(app.config["CONCEPTS_DIR"],
                                    load=app.config['PRELOAD'])
        joblib.dump(cd, app.config['LOC_DICT'])
    else:
        cd = gm.load_concept_models(app.config["CONCEPTS_DIR"],
                                    load=app.config['PRELOAD'])
    return cd


def init():
    LOG.info("Loading lemma dictionaries.")
    with open(app.config["IN_KWD_RAW2LEMMA"], "r") as f0:
        app.config["KWD_RAW2LEMMA"] = json.load(f0)
    app.config["KWD_LEMMA2RAW"] = {v: k for k, v in app.config["KWD_RAW2LEMMA"].items()}

    with open(app.config["IN_CAT_RAW2LEMMA"], "r") as f0:
        app.config["CAT_RAW2LEMMA"] = json.load(f0)
    app.config["CAT_LEMMA2RAW"] = {v: k for k, v in app.config["CAT_RAW2LEMMA"].items()}

    with open(app.config["STI_CONFIG"], "r") as f0:
        cfg = yaml.safe_load(f0)
    weights = cfg["weights"]
    app.config["WEIGHTS"] = weights

    LOG.info("Loading hierarchical classifier.")
    cat_clfs = gm.load_category_models(app.config["CATEGORIES_DIR"])
    cd = load_models_wrapper()
    hclf = gm.HierarchicalClassifier(cat_clfs, cd)
    hclf.load_vectorizer(app.config["IN_VECTORIZER"])
    app.config["HIER_CLF"] = hclf
    LOG.info('Ready for requests.')


def classify_sti(texts, cat_threshold, kwd_threshold, no_categories):
    if type(texts) == str:
        texts = [texts]
    features, feature_matrix = app.config["HIER_CLF"].vectorize(
        texts, app.config["WEIGHTS"]
    )
    cat_preds, concept_preds = app.config["HIER_CLF"].predict(
        feature_matrix, cat_threshold, kwd_threshold, no_categories
    )
    cat_records = [
        [
            {
                "keyword": t[0],
                "probability": float(t[1]),
                "unstemmed": app.config["CAT_LEMMA2RAW"][t[0]],
            }
            for t in record_preds
        ]
        for record_preds in cat_preds
    ]
    concept_records = [
        [
            {
                "keyword": t[0],
                "probability": float(t[1]),
                "unstemmed": app.config["KWD_LEMMA2RAW"][t[0]],
            }
            for t in record_preds
        ]
        for record_preds in concept_preds
    ]
    if type(texts) == str:  # only return one if passed string instead of list
        LOG.info("Single example passed, returning single list.")
        features = features[0]
        cat_records = cat_records[0]
        concept_records = concept_records[0]
    return features, cat_records, concept_records


def _abort(code, msg, usage=True):
    if usage:
        msg += " " + _find_terms_usage() + "'"

    response = jsonify(service_version=SERVICE_VERSION, msg=msg)
    response.status_code = code
    return response


def _find_terms_usage(joinstr="', '", prestr="'"):
    return (
        "Allowed request fields for "
        + FIND_TERMS_METHOD_NAME
        + " method are "
        + prestr
        + joinstr.join(ALLOWED_FIND_TERMS_FIELDS)
    )


def _validate(data):
    for key in data:
        if key not in ALLOWED_FIND_TERMS_FIELDS:
            LOG.warning(f'"{key}" not in {ALLOWED_FIND_TERMS_FIELDS}')
            return False
    return True


# Load all of the data here.
init()


@app.route(f"{SERVICE_ROOT_PATH}/")
def home():
    LOG.info(f"Loading page: {request.host}/{request.path}")
    if "HIER_CLF" not in app.config:
        init()
    return render_template(
        "home.html",
        version=SERVICE_VERSION,
        interface_version=INTERFACE_VERSION,
        root=SERVICE_ROOT_PATH,
        service_url=f"{request.host}/{request.path}",
        methodname=FIND_TERMS_METHOD_NAME,
        usage=_find_terms_usage(joinstr='", "', prestr='"') + '"',
    )


@app.route(f"{SERVICE_ROOT_PATH}/{FIND_TERMS_METHOD_NAME}/", methods=["POST"])
def find_terms():
    if "HIER_CLF" not in app.config:
        init()
    try:
        LOG.debug("Requesting information.")
        data = request.get_json(force=True)
        LOG.debug(f"Data received {str(data)}.")

        if not _validate(data):
            return _abort(400, "Bad Request incorrect field passed.")
        if "text" not in data:
            return _abort(400, "NO passed text field ..._abort.")
        else:
            text = data.get("text")
            if float(data.get("topic_threshold")) == 1.0:
                no_cats = True
            else:
                no_cats = False
            params = {
                "proba_lim": float(data.get("probability_threshold")),
                "topic_threshold": float(data.get("topic_threshold")),
                "request_id": str(data.get("request_id")),
                "no_categories": no_cats,
            }
        LOG.info(f'probability limit: {params["proba_lim"]}')
        LOG.info(f'request_id: {params["request_id"]}')
        features, cat_records, concept_records = classify_sti(
            text,
            params["topic_threshold"],
            params["proba_lim"],
            params["no_categories"],
        )
        payload = {
            "sti_keywords": concept_records,
            "features": features,
            "topic_probabilities": cat_records,
            "request_id": params["request_id"],
            "probability_threshold": params["proba_lim"],
            "topic_threshold": params["topic_threshold"],
        }
        LOG.info("Request Complete")
        return jsonify(
            status="okay",
            code=200,
            messages=[],
            service_version=SERVICE_VERSION,
            interface_version=INTERFACE_VERSION,
            payload=payload,
        )

    except Exception as ex:
        etype = type(ex)
        print(str(ex))

        if etype == ValueError or "BadRequest" in str(etype):
            return _abort(400, str(ex) + ".")
        else:
            print("Service Exception. Msg: " + str(type(ex)))
            return _abort(500, "Internal Service Error", usage=False)


if __name__ == "__main__":
    port = os.getenv("PORT")
    if port is None:
        LOG.info("Setting port to default value of 5000.")
        port = 5000
    app.run(
        host="0.0.0.0", port=int(port), threaded=True,
    )
